image/rhel/konflux.Dockerfile

```diff
@@ -1,7 +1,28 @@
+FROM registry.access.redhat.com/ubi8/nodejs-18:latest AS ui-builder
+
+# Switch to root because ubi8/nodejs image runs as non-root by default which does not let e.g. install yarn.
+USER 0:0
+
+WORKDIR /go/src/github.com/stackrox/rox/app
+
+COPY . .
+
+# Install yarn
+RUN make -C ui/.yarn-dist install
+
+# This sets branding during UI build time. This is to make sure UI is branded as commercial RHACS (not StackRox).
+# ROX_PRODUCT_BRANDING is also set in the resulting image so that Central Go code knows its RHACS.
+# TODO(ROX-22338): switch branding to RHACS_BRANDING when intermediate Konflux repos aren't public.
+ENV ROX_PRODUCT_BRANDING="STACKROX_BRANDING"
+
+# UI build is not hermetic because Cachi2 does not support pulling packages according to V1 of yarn.lock.
+# TODO(ROX-20723): enable yarn package prefetch and make UI builds hermetic.
+RUN make -C ui build
```

Comment:
Edit:

Conceptually the same but probably cleaner and should avoid the issues mentioned above:
```sh
cd /tmp
npm install --no-save yarn@1.22.21
export PATH=/tmp/node_modules/yarn/bin:$PATH
```

Reply:
I had a thing similar to what you describe in [this PR](https://github.com/stackrox/stackrox/pull/8689/commits/c9fe793f372a2fc0f3f79f9c906c694acdabfa3d). There are `package.json`+`package-lock.json` with only `yarn` and yarn is fetched using `npm` package manager [support](https://github.com/containerbuildsystem/cachi2/blob/main/README.md#npm) (search for `prefetch-input` in that commit I linked).

Since that setup is in a separate directory, we don't need to move files around.

It works and I had no issues with its functioning. In fact, colleagues from ACM used a similar approach to create their builder image for downstream use ([one](https://github.com/stolostron/image-builder/blob/master/yarn-source/package.json), [two](https://gitlab.cee.redhat.com/acm-cpaas-midstream/builders-and-bases/-/blob/rhacm-2.10-rhel-9/distgit/containers/acm-yarn-builder-nodejs-20/Dockerfile.in#L19)).

I'm not attached to what I currently have in this PR and can change to that suggestion.

I selected the current thing because:
* It uses something that's native to nodejs - Corepack, i.e. does not require knowledge of Red Hat tools (in this case Cachi2).
* It has fewer moving parts (subjective, but benefit).
* It is similar to providing distribution of `gradle` directly in the repo ([here](https://github.com/stackrox/stackrox/tree/master/qa-tests-backend/gradle/wrapper)) which is a standard practice in Java world.
* Important for me is that it allows us to move away from mentioned ACM builder image downstream as well. Not that I anticipate ACM team to drop their builder image soon, but I think that taking a dependency on publicly available `ubi8/nodejs-18` can be more reliable in longer run, especially as we can switch nodejs versions at our discretion.

It has downsides:
* That tarball is opaque for code reviewing.
* That tarball is 1.2MB. Not terribly lot but still adds.

To counter the first con, I [ran the file through VirusTotal](https://www.virustotal.com/gui/file/490acaf6065bc8ed86beb5dc4f244d92a7f649360c54ce517c8ef39f0a4284a0) to make sure it's clean. Not terribly fancy but gives us some confidence. If we proceed with the current approach, I can modify instructions to make sure we do that each time the file is changed.

Regarding:

> * Not require running the container as root

I don't think this is a huge issue because other containers and their builder stages run as root by default. Running as root is the only way to install rpm packages, for example.
In case of `ubi8/nodejs-18` container, it is the container setting that it's switched to non-root out of the box. I suppose to make it more ready for running workloads. In our case, we use `ubi8/nodejs-18` as one of _builder_ stages. We use it to build UI. Later it gets thrown away after we copy UI assets from it to the final stage (which is set to run as non-root).
Therefore root/non-root thing in UI build is effective only during the container build. Since it's root _in the container_, I don't think it creates more threat, but I may be wrong and there is a practice that I don't know.

----

Regarding

> context of the UI build problems with downstream vs Konflux

There is a big problem of `yarn` `v3`/`v1` compatibility in _prefetching_ UI dependencies that I brought up in Slack <https://redhat-internal.slack.com/archives/C7ERNFL0M/p1708343658894019>. I don't attempt to solve it here (here UI dependencies are fetched from untrusted internet).

The specific thing in this PR is simpler: it's about getting `yarn` executable itself. (Should we have used `yarn` `v3`, we would also have to install it somehow from somewhere.)

Reply:
> I had a thing similar to what you describe in [this PR](https://github.com/stackrox/stackrox/pull/8689/commits/c9fe793f372a2fc0f3f79f9c906c694acdabfa3d).

Aha, yes that would work the same, in a somewhat more rigorous way than the example I posted above.

> I selected the current thing because: ...

IMO these benefits are all obtainable with the approach that downloads `yarn` via `npm` first, except for: "It is similar to providing distribution of gradle directly in the repo", which I don't have a strong opinion on the pros/cons of. We would get the same functionality with a less intrusive change: ~3 lines in the Dockerfile.

> To counter the first con, I ran the file through VirusTotal

I think this is the biggest benefit of your current approach, as my suggested method is an unchecked download via the internet. Do you know if the intention in Konflux is to use a similar tool to scan all dependencies in the cache?

> Since it's root in the container, I don't think it creates more threat, but I may be wrong and there is a practice that I don't know.

This is fair, since this is only an intermediate build stage I doubt there is much security risk. My suggestion was only to remove it since with the local install it would not be necessary.

> There is a big problem of yarn v3/v1 compatibility in prefetching UI dependencies that I brought up in Slack

This was definitely a point of confusion on my part, and your link to [support](https://github.com/containerbuildsystem/cachi2/blob/main/README.md#npm) was a huge help in clarifying why the `npm`/`yarn v1`/`yarn v3` compatibility is such a problem so thanks for the information. My impression was that Konflux/Cachi2 had a dependency on specific package managers versions, but after a browse of the repo it seems like the dependency is more on the package manager _lockfile format_ instead. At least in the case of npm, it looks like they are parsing the package-lock.json, downloading manually into a cachi2-specific-layout, and then using this when the app `npm installs`. 🤔

Reply:
I changed my mind on this and using now the approach with a lockfile for installing `yarn`. I have an idea how to also make it work downstream.

Reply:
Cool, sounds good to me. I'll resolve this comment chain and approve.